compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -4487,11 +4489,20 @@ private boolean removesDistinctRecords(BType typeToRemove, BType remainingType)
             case UNION:
                 for (BType memberType : ((BUnionType) remainingType).getMemberTypes()) {
                     BType referredMemberType = getReferredType(memberType);
-                    if (isClosedRecordTypes(referredMemberType)) {
-                        fieldsInRemainingTypes.add(((BRecordType) referredMemberType).fields.keySet());
+                    int tag = referredMemberType.tag;
+                    if (tag == RECORD) {
+                        BRecordType memberRecordType = (BRecordType) referredMemberType;
+                        if (!memberRecordType.sealed && memberRecordType.restFieldType != symTable.neverType) {
+                            return false;
+                        }
+
+                        fieldsInRemainingTypes.add(memberRecordType.fields.keySet());
```

Comment:
Can we not use the existing logic for Record type in this method itself by recursively calling `removesDistinctRecords`?

Reply:
ah.. didnt notice the `return`